### ISSUE TITLE
API: Expose `_compute_lwork` to facilitate low-level LAPACK use

### ISSUE THREAD

**Issue by larsoner** [scipy/scipy#10575]
I recently [updated a code base to make use of low-level BLAS and LAPACK functions](https://github.com/mne-tools/mne-python/pull/6603) to speed things up. In particular I found that on SciPy `master` doing thousands of small (e.g., 3x3) decompositions spent about 1/3 of the time in `get_lapack_funcs`, 1/3 of the time in `compute_lwork`, and 1/3 of the time actually doing the computation. Getting the LAPACK functions once and computing lwork once cut down on a lot of overhead.

@pv has already tackled making `get_lapack_funcs` faster in #10564, and `_compute_lwork` in #10574, which should help.

However, I propose that we also make `_compute_lwork` public, as it would facilitate use of the low-level LAPACK functions for expert users. Any objections to this?

There are probably other places we could facilitate using low-level functions (e.g., a unified `check_info` function), but already having `_compute_lwork` public would help.

**Comment by sturlamolden** [comment]
+1

Alas, not really needed, we compute the lwork with a special call to the lapack function.

**Comment by sturlamolden** [comment]
If speed is paramount, I recommend to call LAPACK directly in Cython. We expose all of SciPy's LAPACK to Cython now.

**Comment by pv** [comment]
`_compute_lwork` is mostly needed to work around a bug in old(?) LAPACK versions.
I'm not sure if the problem with returning bad lwork from single-precision routines is fixed in current LAPACK.

**Comment by larsoner** [comment]
> If speed is paramount, I recommend to call LAPACK directly in Cython. We expose all of SciPy's LAPACK to Cython now.

Yes this would probably be faster. But for libraries that want to stay in pure Python, making `compute_lwork` public would allow them to get faster for the use cases in the top comment (even if they are still not as fast as Cython would be), so there are merits to this in-between solution.

> _compute_lwork is mostly needed to work around a bug in old(?) LAPACK versions.
I'm not sure if the problem with returning bad lwork from single-precision routines is fixed in current LAPACK.

Based on gh-8670 we require >= 3.4.0, and an [old, nicely documented commit message](https://github.com/scipy/scipy/commit/6aff5830fe18a24979376416740f50a7bd9edbd4) says that the bug exists in (at least) 3.5.0.

FWIW the function also does two other things, which admittedly could be reimplemented in other libraries that use low-level LAPACK functions:

1. Checks the return value for validity (non-zeroness)
2. Triages return value based on the `func` call's return (2 elements or 1)

**Comment by ilayn** [comment]
We have bumped the minimum required lapack version to 3.7.1 now. And I think that bug is fixed already but cannot find it yet in the lapack release notes.
